src/agl/web_app_manager_service_agl.cc

```diff
@@ -292,42 +262,61 @@ bool WebAppManagerServiceAGL::StartService() {
     }
   }
 
-  TriggerStartupApp();
-
   return true;
 }
 
-void WebAppManagerServiceAGL::TriggerStartupApp() {
-  LOG_DEBUG("Triggering app start: %s", startup_app_uri_.c_str());
-  if (!startup_app_uri_.empty()) {
-    if (startup_app_uri_.find("http://") == 0) {
-      startup_app_timer_.Start(
-          10, this, &WebAppManagerServiceAGL::LaunchStartupAppFromURL);
-    } else {
-      startup_app_timer_.Start(
-          10, this, &WebAppManagerServiceAGL::LaunchStartupAppFromJsonConfig);
-    }
-  }
+void WebAppManagerServiceAGL::LaunchOnIdle(const std::string& id,
+                                           const std::string& uri,
+                                           int surface_id,
+                                           AglShellSurfaceType surface_type,
+                                           AglShellPanelEdge panel_type,
+                                           int width,
+                                           int height) {
+  auto launch_data = std::make_unique<LaunchData>();
+  launch_data->app_id = id;
+  launch_data->uri = uri;
+  launch_data->surface_id = surface_id;
+  launch_data->surface_type = surface_type;
+  launch_data->panel_type = panel_type;
+  launch_data->width = width;
+  launch_data->height = height;
+
+  auto* timer = new OneShotTimerWithData<WebAppManagerServiceAGL, LaunchData>();
+  timer->Start(0, this, &WebAppManagerServiceAGL::OnLaunchApp,
+               std::move(launch_data));
 }
 
-void WebAppManagerServiceAGL::TriggetEventForApp(const std::string& action) {
-  if (app_id_event_target_.empty())
-    return;
-
-  if (action == kActivateEvent) {
-    startup_app_timer_.Start(10, this,
-                             &WebAppManagerServiceAGL::OnActivateEvent);
-  } else if (action == kDeactivateEvent) {
-    startup_app_timer_.Start(10, this,
-                             &WebAppManagerServiceAGL::OnDeactivateEvent);
-  } else if (action == kKilledApp) {
-    startup_app_timer_.Start(1000, this, &WebAppManagerServiceAGL::OnKillEvent);
+void WebAppManagerServiceAGL::SendEventOnIdle(const std::string& event,
+                                              const std::string& app_id) {
+  auto event_data = std::make_unique<EventData>();
+  event_data->app_id = app_id;
+  auto* timer = new OneShotTimerWithData<WebAppManagerServiceAGL, EventData>();
+  if (event == kActivateEvent)
+    timer->Start(0, this, &WebAppManagerServiceAGL::OnActivateEvent,
+                 std::move(event_data));
+  else if (event == kDeactivateEvent)
+    timer->Start(0, this, &WebAppManagerServiceAGL::OnDeactivateEvent,
+                 std::move(event_data));
+  else if (event == kKilledApp)
+    timer->Start(1000, this, &WebAppManagerServiceAGL::OnKillEvent,
```

Comment:
So I gather that these are copy-pasted from the previous values. But I wonder, we still keep a 1 second delay for killing it, yet the others (activation/deactivaton) are done instantly. 

And interestingly we move from a 10 ms to a 0 (hence I suppose the mutex is the one handling access). Now given that have serialized access to the "resources" does it really make sense to actually have a timer for this? I mean what purpose would it now serve? 

And why do we wait for for a 1 before killing it?

Reply:
The key here is where the code is running. Even for a 0ms wait it is going to be queued in the main thread event loop. So it is not exactly done "instantly". It is just how we put that task in the main thread.

About why kill delays for 1 second... I think it would likely deserve a comment there. The reason is simply that it allows one second from close to kill request, that helps with synchronizing cookies, etc, to disk.

In any case, in this case it is just the same wait we had. For the 10ms cases those were because IPC was not good. 1s wait is because of dumping to disk after close.

Reality: we don't really have an active application manager killing apps when they are not used anymore. So kill call is not really happening. But once we have such a thing, that's how it should work.

Reply:
Just that it's a bit odd to use a timer to queue tasks, I understand the reasoning for it, but it would have self explanatory if instead of Timer->Start(0); it would have been a QueueTask(Task_Name) or something like that. 

To me a comment for the 0 timeout seems also seems appropriate. // we basically fake a queueing mechanism with a 0 timeout delay. 

Wrt to the kill event, not really sure how these really work together. I was under the impression that the main thread waits for the delay, and then sends the event (timeout occurred). It kind of implies there's another event that was sent in the past; sort of close event, followed by at a later point in time, by the kill event and these two work together, for a clean stop/shutdown of the app. 


> Reality: we don't really have an active application manager killing apps when they are not used anymore

Well, not actively, but you can do it over the cmd line. With the work done by Scott to migrate to systemd user services it would do it using systemd.